hcl2template/types.datasource.go

```diff
@@ -0,0 +1,154 @@
+package hcl2template
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	packersdk "github.com/hashicorp/packer-plugin-sdk/packer"
+	"github.com/hashicorp/packer/packer"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// DataBlock references an HCL 'data' block.
+type Datasource struct {
+	Type string
+	Name string
+
+	value cty.Value
+	block *hcl.Block
+}
+
+type Datasources map[DatasourceRef]Datasource
+
+func (data *Datasource) Ref() DatasourceRef {
+	return DatasourceRef{
+		Type: data.Type,
+		Name: data.Name,
+	}
+}
+
+func (ds *Datasources) Values() (map[string]cty.Value, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+	res := map[string]cty.Value{}
+	valuesMap := map[string]map[string]cty.Value{}
+
+	for ref, datasource := range *ds {
+		if datasource.value == (cty.Value{}) {
+			diags = append(diags, &hcl.Diagnostic{
+				Summary:  fmt.Sprintf("empty value"),
+				Subject:  &datasource.block.DefRange,
+				Severity: hcl.DiagError,
+			})
+			continue
+		}
+		inner := valuesMap[ref.Type]
+		if inner == nil {
+			inner = map[string]cty.Value{}
+		}
+		inner[ref.Name] = datasource.value
+		res[ref.Type] = cty.MapVal(inner)
+
+		// Keeps values of different datasources from same type
+		valuesMap[ref.Type] = inner
+	}
+
+	return res, diags
+}
+
+type DatasourceRef struct {
+	Type string
+	Name string
+}
+
+// the 'addition' field makes of ref a different entry in the data sources map, so
+// Ref is here to make sure only one is returned.
+func (r *DatasourceRef) Ref() DatasourceRef {
+	return DatasourceRef{
+		Type: r.Type,
+		Name: r.Name,
+	}
+}
```

Comment:
I think that if we plan on never adding any other field to a DatasourceRef then, this func is not needed. I had added it to the SourceRef blocks in because I needed to add a body in case the config overloaded the source in a build block, and then if that field was set, then the `SourceRef` would not match in the 'unique source map'.
```suggestion
```

Reply:
This will be removed in there: https://github.com/hashicorp/packer/pull/10304/files#diff-83936f835c1a77350b68035469f13800efca160475d977415a79ce1f17853670

Reply:
( I added an other intermediary type in that could contain the additional block )

Reply:
This ref is being used to check for duplicated data blocks and it makes it very simple by using the ref as a key to look for an existing data source in the map. I see that you still use it to do the same with source blocks. If I remove this I will have to either make the key of the map a string of type+name or make it a list instead and perform a search over a slice to look for duplications.

Adding some sort of `DatasourceUseBlock` will add one unnecessary layer to it because data sources don't have usage specifics like sources. 

Do you have any specific concerns about the DatasourceRef? I think right now it exists only to make it easier to deal with the list of data sources.

Reply:
no concerns, just that this function is not necessary, a `*DatasourceRef` should not be used as a map accessor, but a `DatasourceRef` can, the map type will know what to do. This function is handy because it returns a `DatasourceRef`.

Reply:
I made a suggestion here: https://github.com/hashicorp/packer/pull/10440#discussion_r559584486

Reply:
I'm looking at your PR now and I can also make it as you did with the RequiredPlugins 🤔

Reply:
(DatasourceUseBlock will not be necessary here as we never override this like for a source block)

Reply:
I just realized that this function wasn't being used anywhere. I think I created it by mistake. But there's still the `(data *Datasource) Ref() DatasourceRef` which is the one being actually used.

I'm not using `*DatasourceRef` as the map accessor 🤔

Reply:
ah 🙌🏼  good catch !